### ISSUE TITLE
Nemesis writing/reading Issue

### ISSUE THREAD

**Issue by friedmud** [libMesh/libmesh#1087]
It seems as if either our Nemesis writing or our Nemesis reading is buggered.  Using NemesisIO to write out a Mesh and then reading it directly back in results in errors.

Here is a minimal libMesh code to demonstrate the issue and I'll attach a mesh to go with it.  For the attached mesh this application works on 2 processors but fails on 4.

@jwpeterson Can I get some help here?

Mesh: [pin_cell_fine.e.zip](https://github.com/libMesh/libmesh/files/467530/pin_cell_fine.e.zip)

```C++
#include "libmesh/libmesh.h"
#include "libmesh/mesh.h"
#include "libmesh/nemesis_io.h"
#include "libmesh/metis_partitioner.h"
#include "libmesh/parallel_mesh.h"

using namespace libMesh;

int main (int argc, char ** argv)
{
  LibMeshInit init (argc, argv);

  if (libMesh::on_command_line("--help"))
  {
    std::cout<<"Example: ./simple-opt --mesh filename.e"
             <<"--mesh             Full name of the mesh file to read in. \n"
             <<std::endl;

    return 0;
  }

  std::string filename = libMesh::command_line_next("--mesh", std::string());
  unsigned int n_procs = libMesh::command_line_next("--n-procs", 1);

  Parallel::Communicator & comm = init.comm();

  Mesh mesh(init.comm());

  mesh.read(filename);

  MetisPartitioner partitioner;

  partitioner.partition(mesh);

  {
    Nemesis_IO nemesis(mesh);
    nemesis.verbose(true);
    nemesis.write(filename);
  }

  {
    DistributedMesh reader_mesh(init.comm());
    Nemesis_IO reader(reader_mesh);
    reader.verbose(true);
    reader.read(filename);
  }

  return 0;
}
```

Here is the error message:

```
Exodus Library Error: [ex_get_node_cmap]
	Error: failed to get variable "n_comm_nids" from file ID 65536
	[-40] Index exceeds dimension bound
Error reading node cmap node and processor ids!
[0] src/mesh/nemesis_io_helper.C, line 393, compiled nodate at notime
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```

**Comment by jwpeterson** [comment]
On Mon, Sep 12, 2016 at 8:33 AM, Derek Gaston <notifications@github.com>
wrote:

> It seems as if either our Nemesis writing or our Nemesis reading is
> buggered. Using NemesisIO to write out a Mesh and then reading it directly
> back in results in errors.
>
> Here is a minimal libMesh code to demonstrate the issue and I'll attach a
> mesh to go with it. For the attached mesh this application works on 2
> processors but fails on 4.
>
> @jwpeterson <https://github.com/jwpeterson> Can I get some help here?
>
> Mesh: pin_cell_fine.e.zip
> <https://github.com/libMesh/libmesh/files/467530/pin_cell_fine.e.zip>
>
> #include "libmesh/libmesh.h"
> #include "libmesh/mesh.h"
> #include "libmesh/nemesis_io.h"
> #include "libmesh/metis_partitioner.h"
> #include "libmesh/parallel_mesh.h"
> using namespace libMesh;
> int main (int argc, char ** argv)
> {
>   LibMeshInit init (argc, argv);
>
>   if (libMesh::on_command_line("--help"))
>   {
>     std::cout<<"Example: ./simple-opt --mesh filename.e"
>              <<"--mesh             Full name of the mesh file to read in. \n"
>              <<std::endl;
>
>     return 0;
>   }
>
>   std::string filename = libMesh::command_line_next("--mesh", std::string());
>   unsigned int n_procs = libMesh::command_line_next("--n-procs", 1);
>
>   Parallel::Communicator & comm = init.comm();
>
>   Mesh mesh(init.comm());
>
>   mesh.read(filename);
>
>   MetisPartitioner partitioner;
>
>   partitioner.partition(mesh);
>
>   {
>     Nemesis_IO nemesis(mesh);
>     nemesis.verbose(true);
>     nemesis.write(filename);
>   }
>
>   {
>     DistributedMesh reader_mesh(init.comm());
>     Nemesis_IO reader(reader_mesh);
>     reader.verbose(true);
>     reader.read(filename);
>   }
>
>   return 0;
> }
>
> Here is the error message:
>
> Exodus Library Error: [ex_get_node_cmap]
>     Error: failed to get variable "n_comm_nids" from file ID 65536
>     [-40] Index exceeds dimension bound
> Error reading node cmap node and processor ids!
> [0] src/mesh/nemesis_io_helper.C, line 393, compiled nodate at notime
> application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
>
>
I will take a look.

I'm kind of surprised it isn't working... we have a test of reading Nemesis
files in MOOSE: tests/mesh/nemesis/nemesis_test.i

Can you think of anything different that we do in MOOSE to make this work?
Does your attached test code work with the cylinder mesh from that MOOSE
test in place of pin_cell_fine?

-- 
John

**Comment by friedmud** [comment]
That test only tests _reading_ Nemesis... not writing it then reading it back in.  I assume that those `cylinder` mesh files were generated using the old Nemesis "split" script using the old tools... which is why they work.  It could also be that we're just getting lucking with that particular mesh (the one I posted above _does_ work on 2 processors... just not on 4.  Others I've tested have worked/failed on basically random numbers of processors).

I expect that the problem here is actually in the _writing_ of the Nemesis files.  But it must be some data in the writing that, for instance, Paraview doesn't care about because Paraview reads our Nemesis results just fine (including being able to read the files that libMesh can't read back in from the code above).

**Comment by jwpeterson** [comment]
> I assume that those cylinder mesh files were generated using the old Nemesis "split" script using the old tools

Must have been from that loadbal/nem_spread script originally, yeah.  Another possibility: Metis sometimes gives non-contiguous partitionings, and the Nemesis writer doesn't handle that case well...?  I'll let you know what I find.

**Comment by jwpeterson** [comment]
@friedmud, if I understand correctly, I should *not* need to have your PR branch checked to run this test code, correct?  Therefore, I just run it in parallel to determine the number of output files...

It currently gives me an error on 1, 2, and 4 procs (the error on 4 procs matches the one you reported). Am I missing something?

1 processor:
```
Exodus Library Error: [ex_update]
	Error: failed to update file id 0
	[-33] Not a netcdf id
Error flushing buffers to file.
```

2 processors:
```
Assertion `this->comm().semiverify(obj ? &uniqueid : nullptr)' failed.
```

4 processors:
```
Exodus Library Error: [ex_get_node_cmap]
	Error: failed to get variable "n_comm_nids" from file ID 65536
	[-40] Index exceeds dimension bound
Error reading node cmap node and processor ids!
[0] ../src/mesh/nemesis_io_helper.C, line 378, compiled Sep 12 2016 at 12:27:40
```

**Comment by jwpeterson** [comment]
BTW, I think the error in the 1 processor case is due to the fact that we call `exII::ex_update()` in `~Nemesis_IO_Helper()` on a file that is open for *reading*.  According to the Exodus docs, `ex_update()` is only for files that are open for *writing*... this one is fixed on my fork in 827bac3.

**Comment by friedmud** [comment]
Right - this should work with any libMesh... just run it with different MPI.

Not sure why it's working for me on 2 procs... but at least you have a test
case :-)

Thanks for looking into this!!

**Comment by jwpeterson** [comment]
The assert failure on 2 procs was traced down to a bad MPI type declaration.  This is fixed in dc68637, but it isn't a complete fix since we need to throw an error somewhere in the MPI-1 case.  @roystgnr where do you think the best place for that is?

**Comment by friedmud** [comment]
Awesome @jwpeterson ! I really appreciate it!

**Comment by jwpeterson** [comment]
> On Sep 12, 2016, at 4:59 PM, Derek Gaston <notifications@github.com> wrote:
> 
> Awesome @jwpeterson ! I really appreciate it!

The actual problem isn't fixed yet unfortunately... but these two needed to be fixed as well.

**Comment by friedmud** [comment]
Not a problem - I just really appreciate you working on it!

**Comment by jwpeterson** [comment]
#1093 fixes the bug for me on four procs, but I see several other places in the code where this same pattern is used... and should probably be avoided.  This is also probably going to conflict with your #1086, but the resolution should be fairly straightforward.

**Comment by friedmud** [comment]
#1086 needs some work anyway.  I want to make it so that you can do the partitioning in parallel (like use 12 processors to write out 1,000 Nemesis files).  My first attempt at that won't work because there is a bit more parallel communication in NemesisIO_Helper where it's trying to get information about the global number of blocks and sidesets / etc.  I have an idea about how to fix that...

Thanks for working on this - I'll give it a try!

**Comment by friedmud** [comment]
This was all looking really good.  I was able to partition a file up into 48 chunks and run a medium sized 2D problem on Falcon... but then going the next step did not work out.

Currently getting a segfault with a fairly large 3D problem on 96 procs... bummer.

Building a dgb executable now... hopefully I can get a stack trace...

**Comment by roystgnr** [comment]
@friedmud, are you using adaptivity too by any chance?  There's currently two bugs with DistributedMesh AMR that I'm tracking down.  Both are extremely hard to replicate and are going to be harder to fix, but if they might be affecting you then there's a workaround I could put in for the meantime.

**Comment by friedmud** [comment]
Well - I let it run in dbg mode for a few hours on 96 procs... and it didn't make it to the segfault.

I'll try to generate a minimal example today...

**Comment by friedmud** [comment]
Sorry @roystgnr , didn't see your message.

Nope: I'm not using adaptivity (yet!).  I _definitely_ want to do that in the next couple of months though... so I'm happy to hear that you're working out problems there!

**Comment by friedmud** [comment]
Well... that's frustrating... can't seem to reproduce it on a smaller problem.  I'll keep trying though... you definitely don't want to debug with my 10M element mesh...

**Comment by friedmud** [comment]
Ok - I can trigger an issue on 12 processors... it hangs when trying to read the mesh.  Here is a backtrace from one of the processes.

```
* thread #1: tid = 0x841274d, 0x00000001108ff018 libpmpi.12.dylib`MPID_nem_mpich_blocking_recv + 600, stop reason = signal SIGSTOP
  * frame #0: 0x00000001108ff018 libpmpi.12.dylib`MPID_nem_mpich_blocking_recv + 600
    frame #1: 0x00000001108fe5cf libpmpi.12.dylib`MPIDI_CH3I_Progress + 399
    frame #2: 0x000000011086ea29 libpmpi.12.dylib`MPIC_Wait + 105
    frame #3: 0x000000011086f03e libpmpi.12.dylib`MPIC_Recv + 510
    frame #4: 0x000000011073639b libpmpi.12.dylib`MPIR_Bcast_binomial + 1355
    frame #5: 0x0000000110734cf9 libpmpi.12.dylib`MPIR_Bcast_intra + 825
    frame #6: 0x000000011073866c libpmpi.12.dylib`MPIR_Bcast + 76
    frame #7: 0x00000001107385b2 libpmpi.12.dylib`MPIR_Bcast_impl + 226
    frame #8: 0x000000011072f167 libpmpi.12.dylib`MPIR_Allreduce_intra + 1319
    frame #9: 0x00000001107317fb libpmpi.12.dylib`MPIR_Allreduce_impl + 267
    frame #10: 0x0000000110601657 libmpi.12.dylib`MPI_Allreduce + 3799
    frame #11: 0x000000010d5dedfd libmesh_dbg.0.dylib`void libMesh::Parallel::Communicator::min<unsigned int>(this=0x00007fcd63b08dc0, r=0x00007fff54de7d24) const + 205 at parallel_implementation.h:1541
    frame #12: 0x000000010d7955d4 libmesh_dbg.0.dylib`bool libMesh::Parallel::Communicator::semiverify<unsigned int>(this=0x00007fcd63b08dc0, r=0x0000000000000000) const + 100 at parallel_implementation.h:1410
    frame #13: 0x000000010d806972 libmesh_dbg.0.dylib`libMesh::DistributedMesh::libmesh_assert_valid_parallel_flags(this=0x00007fcd63cb2060) const + 1810 at parallel_mesh.C:964
    frame #14: 0x000000010d809916 libmesh_dbg.0.dylib`libMesh::DistributedMesh::delete_remote_elements(this=0x00007fcd63cb2060) + 102 at parallel_mesh.C:1359
    frame #15: 0x000000010d7c8074 libmesh_dbg.0.dylib`libMesh::Nemesis_IO::read(this=0x00007fff54dec058, base_filename="../../geom/assemblies/31enr_no_instr_15nw_extruder_in.e") + 53476 at nemesis_io.C:1168
    frame #16: 0x000000010bd923dd libmoose-dbg.0.dylib`FileMesh::buildMesh(this=0x00007fcd66012400) + 429 at FileMesh.C:68
    frame #17: 0x000000010bddc62d libmoose-dbg.0.dylib`MooseMesh::init(this=0x00007fcd66012400) + 11293 at MooseMesh.C:1833
    frame #18: 0x000000010b492075 libmoose-dbg.0.dylib`SetupMeshAction::act(this=0x00007fcd64020c00) + 2053 at SetupMeshAction.C:148
    frame #19: 0x000000010b3e73d2 libmoose-dbg.0.dylib`ActionWarehouse::executeActionsWithAction(this=0x00007fcd640099b0, task="init_mesh") + 1138 at ActionWarehouse.C:325
    frame #20: 0x000000010b3e6f2d libmoose-dbg.0.dylib`ActionWarehouse::executeAllActions(this=0x00007fcd640099b0) + 509 at ActionWarehouse.C:302
    frame #21: 0x000000010ba0dbd5 libmoose-dbg.0.dylib`MooseApp::runInputFile(this=0x00007fcd64009400) + 1013 at MooseApp.C:369
    frame #22: 0x000000010ba111e5 libmoose-dbg.0.dylib`MooseApp::run(this=0x00007fcd64009400) + 101 at MooseApp.C:575
    frame #23: 0x000000010ae11b1f mocodile-dbg`main(argc=3, argv=0x00007fff54deee80) + 303 at main.C:26
    frame #24: 0x00007fff9fa335ad libdyld.dylib`start + 1
```

**Comment by friedmud** [comment]
(BTW: For this mesh - it worked fine on 4 and 6 processors)

**Comment by friedmud** [comment]
Looks like some are doing `min()` while others are doing `max()`?

```
* thread #1: tid = 0x841274b, 0x000000011199bf16 libpmpi.12.dylib`MPID_nem_mpich_blocking_recv + 342, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x000000011199bf16 libpmpi.12.dylib`MPID_nem_mpich_blocking_recv + 342
    frame #1: 0x000000011199b5cf libpmpi.12.dylib`MPIDI_CH3I_Progress + 399
    frame #2: 0x000000011190ba29 libpmpi.12.dylib`MPIC_Wait + 105
    frame #3: 0x000000011190c03e libpmpi.12.dylib`MPIC_Recv + 510
    frame #4: 0x00000001117d339b libpmpi.12.dylib`MPIR_Bcast_binomial + 1355
    frame #5: 0x00000001117d1cf9 libpmpi.12.dylib`MPIR_Bcast_intra + 825
    frame #6: 0x00000001117d566c libpmpi.12.dylib`MPIR_Bcast + 76
    frame #7: 0x00000001117d55b2 libpmpi.12.dylib`MPIR_Bcast_impl + 226
    frame #8: 0x00000001117cc167 libpmpi.12.dylib`MPIR_Allreduce_intra + 1319
    frame #9: 0x00000001117ce7fb libpmpi.12.dylib`MPIR_Allreduce_impl + 267
    frame #10: 0x000000011169a657 libmpi.12.dylib`MPI_Allreduce + 3799
    frame #11: 0x000000010dff6c4f libmesh_dbg.0.dylib`void libMesh::Parallel::Communicator::max<unsigned int>(this=0x00007fcad8c0bc90, r=0x00007fff53d3fd20) const + 191 at parallel_implementation.h:1730
    frame #12: 0x000000010e82b5e1 libmesh_dbg.0.dylib`bool libMesh::Parallel::Communicator::semiverify<unsigned int>(this=0x00007fcad8c0bc90, r=0x0000000000000000) const + 113 at parallel_implementation.h:1411
    frame #13: 0x000000010e89cace libmesh_dbg.0.dylib`libMesh::DistributedMesh::libmesh_assert_valid_parallel_flags(this=0x00007fcad8ca35a0) const + 2158 at parallel_mesh.C:968
    frame #14: 0x000000010e89f916 libmesh_dbg.0.dylib`libMesh::DistributedMesh::delete_remote_elements(this=0x00007fcad8ca35a0) + 102 at parallel_mesh.C:1359
    frame #15: 0x000000010e85e074 libmesh_dbg.0.dylib`libMesh::Nemesis_IO::read(this=0x00007fff53d44058, base_filename="../../geom/assemblies/31enr_no_instr_15nw_extruder_in.e") + 53476 at nemesis_io.C:1168
    frame #16: 0x000000010ce283dd libmoose-dbg.0.dylib`FileMesh::buildMesh(this=0x00007fcad9816400) + 429 at FileMesh.C:68
    frame #17: 0x000000010ce7262d libmoose-dbg.0.dylib`MooseMesh::init(this=0x00007fcad9816400) + 11293 at MooseMesh.C:1833
    frame #18: 0x000000010c528075 libmoose-dbg.0.dylib`SetupMeshAction::act(this=0x00007fcadc80ec00) + 2053 at SetupMeshAction.C:148
    frame #19: 0x000000010c47d3d2 libmoose-dbg.0.dylib`ActionWarehouse::executeActionsWithAction(this=0x00007fcad98083b0, task="init_mesh") + 1138 at ActionWarehouse.C:325
    frame #20: 0x000000010c47cf2d libmoose-dbg.0.dylib`ActionWarehouse::executeAllActions(this=0x00007fcad98083b0) + 509 at ActionWarehouse.C:302
    frame #21: 0x000000010caa3bd5 libmoose-dbg.0.dylib`MooseApp::runInputFile(this=0x00007fcad9807e00) + 1013 at MooseApp.C:369
    frame #22: 0x000000010caa71e5 libmoose-dbg.0.dylib`MooseApp::run(this=0x00007fcad9807e00) + 101 at MooseApp.C:575
    frame #23: 0x000000010beb9b1f mocodile-dbg`main(argc=3, argv=0x00007fff53d46e80) + 303 at main.C:26
    frame #24: 0x00007fff9fa335ad libdyld.dylib`start + 1
```

**Comment by friedmud** [comment]
Ok - I guess this isn't hung... it's just INCREDIBLY slow...

**Comment by roystgnr** [comment]
Yeah, sorry.  The DistributedMesh code paths are currently incredibly larded up with asserts.  Plus, in dbg mode you're getting the GLIBCXX_DEBUG stuff, which changes the asymptotic complexity of a few std::set operations, which you do *not* want to do with ten million elements...

**Comment by friedmud** [comment]
No problem!  Currently working with about 200k elements.

Slowly going through and removing asserts currently.  May just switch to `devel` mode soon (but I like to let the asserts run for a little while... at some point they have to catch _something_!)

**Comment by roystgnr** [comment]
I'd suggest leaving all the asserts but switching to devel mode.  Nearly every assert in there is there because, years ago, some use case with a prehistoric ParallelMesh was making me suspicious at that point.  The GLIBCXX_DEBUG stuff, on the other hand, is much less likely to catch anything, and is much more likely to be slowing you down (by a factor of O(N/log(N))?  I forget the exact assert that breaks asymptotic complexity guarantees...)

You also might want to cherry-pick https://github.com/libMesh/libmesh/pull/1064/commits/53047a16fb1c029f38165527c66aee40be270e87 just to make sure the few asserts which were killing me aren't also affecting you.

**Comment by friedmud** [comment]
Switching to devel got me there faster. 

Here is the error:

```
Assertion `comp < this->n_comp(s,var)' failed.
comp = 0
this->n_comp(s,var) = 0


[9] ./include/libmesh/dof_object.h, line 808, compiled nodate at notime
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 9
Assertion `comp < this->n_comp(s,var)' failed.
comp = 0
this->n_comp(s,var) = 0


[5] ./include/libmesh/dof_object.h, line 808, compiled nodate at notime
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 5
```

I was able to start all 12 processes (sigh) in lldb and get a stack trace.  Looks like the issue is in sparsity pattern construction....

```
* thread #1: tid = 0x8452522, 0x00000001121a0240 libmpi.12.dylib`MPI_Abort, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
  * frame #0: 0x00000001121a0240 libmpi.12.dylib`MPI_Abort
    frame #1: 0x000000010f8f5b11 libmesh_devel.0.dylib`libMesh::libmesh_terminate_handler() + 65 at libmesh.C:322
    frame #2: 0x00007fff9cb0619e libc++abi.dylib`std::__terminate(void (*)()) + 8
    frame #3: 0x00007fff9cb06213 libc++abi.dylib`std::terminate() + 51
    frame #4: 0x000000010f8a4afb libmesh_devel.0.dylib`__clang_call_terminate + 11
    frame #5: 0x000000010f8a2453 libmesh_devel.0.dylib`::.omp_outlined.(.global_tid.=<unavailable>, .bound_tid.=<unavailable>, __context=<unavailable>) + 291 at threads_pthread.h:399
    frame #6: 0x000000010e8c8513 libomp.dylib`__kmp_invoke_microtask + 147
    frame #7: 0x000000010e88ca9a libomp.dylib`__kmp_fork_call + 3562
    frame #8: 0x000000010e883360 libomp.dylib`__kmpc_fork_call + 192
    frame #9: 0x000000010f8a4622 libmesh_devel.0.dylib`void libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, libMesh::SparsityPattern::Build>(range=<unavailable>, body=<unavailable>) + 626 at threads_pthread.h:386
    frame #10: 0x000000010f88e1fe libmesh_devel.0.dylib`libMesh::DofMap::build_sparsity(this=0x00007fb72cd62b60, mesh=0x00007fb72cadc070) const + 254 at dof_map.C:89
    frame #11: 0x000000010f89aa6f libmesh_devel.0.dylib`libMesh::DofMap::compute_sparsity(this=0x00007fb72cd62b60, mesh=<unavailable>) + 31 at dof_map.C:1649
    frame #12: 0x000000010ff63520 libmesh_devel.0.dylib`libMesh::ImplicitSystem::init_matrices(this=0x00007fb72cd62710) + 224 at implicit_system.C:137
    frame #13: 0x000000010ffc0353 libmesh_devel.0.dylib`libMesh::TransientSystem<libMesh::NonlinearImplicitSystem>::init_data(this=0x00007fb72cd62710) + 19 at transient_system.C:109
    frame #14: 0x000000010ff71538 libmesh_devel.0.dylib`libMesh::System::init(this=0x00007fb72cd62710) + 56 at system.C:242
    frame #15: 0x000000010ff3b412 libmesh_devel.0.dylib`libMesh::EquationSystems::init(this=0x00007fb72e804ab8) + 802 at equation_systems.C:121
    frame #16: 0x000000010ec80a75 libmoose-devel.0.dylib`FEProblem::init(this=0x00007fb72e804800) + 709 at FEProblem.C:3191
    frame #17: 0x000000010eb7557a libmoose-devel.0.dylib`ActionWarehouse::executeActionsWithAction(this=0x00007fb72e8005b0, task="init_problem") + 714 at ActionWarehouse.C:325
    frame #18: 0x000000010eb7529b libmoose-devel.0.dylib`ActionWarehouse::executeAllActions(this=0x00007fb72e8005b0) + 123 at ActionWarehouse.C:302
    frame #19: 0x000000010ed57afa libmoose-devel.0.dylib`MooseApp::runInputFile(this=0x00007fb72e800000) + 266 at MooseApp.C:369
    frame #20: 0x000000010ed5903f libmoose-devel.0.dylib`MooseApp::run(this=0x00007fb72e800000) + 95 at MooseApp.C:575
    frame #21: 0x000000010e873d92 mocodile-devel`main(argc=<unavailable>, argv=<unavailable>) + 114 at main.C:26
```

**Comment by friedmud** [comment]
I'm going to see if I can get it to reproduce on an even smaller problem... I suspect that one of those `verify()` functions would probably tell us something is wrong here...

**Comment by roystgnr** [comment]
Thanks!  I'm about to knock off for the day but if you get it on a problem I can replicate this weekend I'd appreciate it.

**Comment by friedmud** [comment]
Cool - I'm going to spend the rest of today trying to narrow it down.  I _really_ need parallel mesh... I'm in a spot where I simply can't continue without it.

Hopefully I'll have something small and reproducible for us to all play with within a few hours...

**Comment by roystgnr** [comment]
What variable FETypes are you using?

**Comment by friedmud** [comment]
One Lagrange and whole bunch of constant monomials (like dozens).

**Comment by friedmud** [comment]
I also have a few Systems... like 4 or so (unless I'm forgetting some)

**Comment by roystgnr** [comment]
use_coupled_neighbor_dofs() is false?

**Comment by friedmud** [comment]
Hmmm.  It might not be.  I can't remember how we control that currently (usually it's automatic based on whether you're using a discontinuous basis function... right?)

**Comment by roystgnr** [comment]
If you're using *all* discontinuous basis functions then we assume you're doing DG and we turn it on.  Even one LAGRANGE in your system should turn it back off, though, unless you turn it on programatically or via command line flag.

**Comment by friedmud** [comment]
Is it per system?  I definitely have Systems that have _only_ monomials.

**Comment by friedmud** [comment]
I just tried using `--implicit_neighbor_dofs 0` and it didn't make any difference...

**Comment by permcody** [comment]
> I really need parallel mesh...  I'm in a spot where I simply can't continue without it.

Well, that's not something I expected to hear out of @friedmud, pretty much ever, but there it is...

**Comment by friedmud** [comment]
lol!

I never said parallel mesh was worthless... just that most people that think they need it... they really don't ;-)

**Comment by friedmud** [comment]
I've got it to fail now with 50k elements in 2D on 9 processors.

That may be the best I can do.  I'm going to pack it all up and post it....

**Comment by friedmud** [comment]
I got sidetracked on another issue in my code last night (using night loosely... I was working on it until 5AM this morning... ugh.)

Back on this now...

**Comment by friedmud** [comment]
Ok - so I haven't been able to reproduce this with my `simple_libmesh_app`... it appears that it might be something that MOOSE is doing with Nemesis/Distributed mesh.  The very same decomposed Nemesis files seem to work just fine.

I've even tried to reproduce the series of events MOOSE takes between reading the mesh and initializing the system (which is when libMesh runs into problems building the sparsity pattern)... but still no dice.

Still working on it... but come Monday I may need some help...

**Comment by friedmud** [comment]
So - I was able to "fix" it by allowing libMesh to repartition (which we currently had disabled in MOOSE)... but I'm not sure if that actually fixes it... or if the new partitioning just doesn't display the bad behavior.

Even with that "fix" I'm still seeing failure with my real application/mesh at 96 processors.  So I think the above PR is just a red herring.

Unfortunately... I'm struggling to create a small test that shows the problem.  Even that problem that is failing at 96 processors runs fine at 48....

**Comment by friedmud** [comment]
This issue is incredibly frustrating... because it's soooo close to working.  And when it does work it's beautiful.

I just got it to work with 52 MPI with my real code/mesh.  Problem started about 10x faster... and used ~2GB of RAM per process instead of ~8GB like it does with Exodus!  Nearly brought tears to my eyes that I can't seem to get it to go on more processors than that...

So... it's definitely related to how small the domains are.

I'm still searching for a reasonably sized problem that exhibits the bad behavior.

**Comment by friedmud** [comment]
(BTW: Bad behavior is a smörgåsbord of segfaulting, memory corruption and trying to allocate TB of RAM and having the OS kill the process.  There's definitely some bad memory being read somehow...)

**Comment by permcody** [comment]
This is intriguing. Hopefully John, Roy and you can figure this out. I'd
like to try it on my problem to so I'm willing to look at it as well
